File: python/problem118.py
class Solution:
    def generate(self, numRows):
        """
        :type numRows: int
        :rtype: List[List[int]]
        """
        if numRows==0:
            return []
        if numRows==1:
            return [[1]]
        if numRows==2:
            return [[1],[1,1]]
        if numRows>2:
            
            li=[[1],[1,1]]
            
            for i in range(2,numRows):
                li.append([])
                li[i].append(1)
                # li.append([1])
                for k in range(1,i):
                    li[i].append(li[i-1][k]+li[i-1][k-1])
                li[i].append(1)
            return li
                  

File: cpp/problem118.cpp
#include "mystd.h"
using namespace std;

/**
 * a trival question and solution
 * Time and space costs: Theta(n^2)
 */
class Solution {
public:
    vector<vector<int>> generate(int numRows) {
    	vector<vector<int> > res;
    	if( ! numRows ) return res;
    	res.emplace_back( 1, 1 );
    	for (int i = 2; i <= numRows; ++i){
    		res.emplace_back( i, 0 );
    		auto &p = res[ i - 2 ], &q = res[ i-1 ];
    		for (int j = 1; j < i - 1; ++j)  q[ j ] = p[ j-1 ] + p[j];
    		q[ 0 ] = q[ i - 1 ] = 1;
    	}
    	return res;
    }
};


void prt( vector<vector<int>> & v ){
	for(auto & i: v){
		for(auto & j: i){
			cout << j << ' ';
		}
		cout << endl;
	}
}

int main(int argc, char const *argv[]){
	Solution sol;
	auto res = sol.generate(5);

	prt( res );
	
	return 0;
}